tests/unit-tests/test-class-sensei-course-enrolment.php

```diff
@@ -0,0 +1,385 @@
+	 */
+	public function testTriggerCourseEnrolmentCheckSimple() {
+		$course_id  = $this->getSimpleCourse();
+		$student_id = $this->createStandardStudent();
+
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Denies_Crooks::class );
+
+		$course_enrolment = Sensei_Course_Enrolment::get_course_instance( $course_id );
+
+		$pre_crook_enrolment = $course_enrolment->is_enroled( $student_id );
+		$this->assertTrue( $pre_crook_enrolment, 'As a non-crook, this student should be enroled' );
+
+		$this->turnStudentIntoCrook( $student_id );
+		$pre_notify_enrolment = $course_enrolment->is_enroled( $student_id );
+		$this->assertTrue( $pre_notify_enrolment, 'Even as a crook, the cached value should be used until we notify the course enrolment handler' );
+
+		$course_enrolment->trigger_course_enrolment_check( $student_id );
+		$post_notify_enrolment = $course_enrolment->is_enroled( $student_id );
+		$this->assertFalse( $post_notify_enrolment, 'Now that the crook status is known, they should no longer be enroled in the course' );
+	}
+
+	/**
+	 * Tests storage of the term association for positive results.
+	 *
+	 * @covers \Sensei_Course_Enrolment::save_enrolment
+	 */
+	public function testSaveEnrolmentStore() {
+		$course_id  = $this->getSimpleCourse();
+		$student_id = $this->createStandardStudent();
+
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Always_Provides::class );
+
+		$course_enrolment = Sensei_Course_Enrolment::get_course_instance( $course_id );
+		$this->assertTrue( $course_enrolment->is_enroled( $student_id ) );
+
+		$student_term = Sensei_Learner::get_learner_term( $student_id );
+		$this->assertTrue( has_term( $student_term->term_id, Sensei_PostTypes::LEARNER_TAXONOMY_NAME, $course_id ), 'Student term was not associated with the course but should have been' );
+	}
+
+	/**
+	 * Tests remove of the stored term association for when enrolment is removed.
+	 *
+	 * @covers \Sensei_Course_Enrolment::save_enrolment
+	 */
+	public function testSaveEnrolmentRemoved() {
+		$course_id  = $this->getSimpleCourse();
+		$student_id = $this->createStandardStudent();
+
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Denies_Crooks::class );
+
+		$course_enrolment = Sensei_Course_Enrolment::get_course_instance( $course_id );
+		$this->assertTrue( $course_enrolment->is_enroled( $student_id ) );
+
+		$student_term = Sensei_Learner::get_learner_term( $student_id );
+		$this->assertTrue( has_term( $student_term->term_id, Sensei_PostTypes::LEARNER_TAXONOMY_NAME, $course_id ), 'Student term was not associated with the course but should have been' );
+
+		// Turns the student into someone who won't be enroled by provider.
+		$this->turnStudentIntoCrook( $student_id );
+		$course_enrolment->trigger_course_enrolment_check( $student_id );
+
+		$this->assertFalse( has_term( $student_term->term_id, Sensei_PostTypes::LEARNER_TAXONOMY_NAME, $course_id ), 'Student term associated was not removed when enrolment was removed' );
```

Comment:
If you agree with the previous comments about to update some assertion description, we could update these latest too :)

Reply:
0c80b60fbf0e34a8e5eb5c2c4124ba00f7021d0e